File: app/src/main/java/com/example/fruitsdiary/data/fruit/FruitRepository.java
package com.example.fruitsdiary.data.fruit;

import com.example.fruitsdiary.model.Fruit;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;

public class FruitRepository {

    private final FruitDataSource mFruitDataSource;

    @Inject
    FruitRepository(FruitDataSource fruitDataSource) {
        mFruitDataSource = fruitDataSource;
    }

    public Observable<List<Fruit>> getFruits() {
        return mFruitDataSource.getFruits();
    }
}

File: app/src/main/java/com/example/fruitsdiary/data/fruit/FruitRepository.kt
package com.example.fruitsdiary.data.fruit

import com.example.fruitsdiary.model.Fruit
import io.reactivex.Completable
import io.reactivex.Flowable
import javax.inject.Inject

class FruitRepository
@Inject constructor(
        private val fruitRemoteDataSource: FruitRemoteDataSource,
        private val fruitLocalDataSource: FruitLocalDataSource) {

    fun getFruits(): Flowable<List<Fruit>> = fruitRemoteDataSource.getFruits()
                .onErrorResumeNext(fruitLocalDataSource.getFruits())

    fun insertAll(fruitList: List<Fruit>): Completable = fruitLocalDataSource.insertAll(fruitList)

}
